src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+use super::super::SinsemillaInstructions;
+use super::{get_s_by_idx, CellValue, EccPoint, SinsemillaChip, Var};
+use crate::{
+    circuit::gadget::utilities::copy,
+    primitives::sinsemilla::{self, lebs2ip_k, INV_TWO_POW_K},
+};
+use halo2::{
+    circuit::{Chip, Region},
+    plonk::Error,
+};
+
+use ff::{Field, PrimeFieldBits};
+use group::Curve;
+use pasta_curves::{
+    arithmetic::{CurveAffine, FieldExt},
+    pallas,
+};
+
+use std::ops::Deref;
+
+impl SinsemillaChip {
+    #[allow(non_snake_case)]
+    #[allow(clippy::type_complexity)]
+    pub(super) fn hash_message(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        Q: pallas::Affine,
+        message: &<Self as SinsemillaInstructions<
+            pallas::Affine,
+            { sinsemilla::K },
+            { sinsemilla::C },
+        >>::Message,
+    ) -> Result<(EccPoint, Vec<Vec<CellValue<pallas::Base>>>), Error> {
+        let config = self.config().clone();
+        let mut offset = 0;
+
+        // Get the `x`- and `y`-coordinates of the starting `Q` base.
+        let x_q = *Q.coordinates().unwrap().x();
+        let y_q = *Q.coordinates().unwrap().y();
+
+        // Initialize the accumulator to `Q`.
+        let (mut x_a, mut y_a): (X<pallas::Base>, Y<pallas::Base>) = {
+            // Constrain the initial x_q to equal the x-coordinate of the domain's `Q`.
+            let fixed_x_q =
+                region.assign_fixed(|| "fixed x_q", config.constants, offset, || Ok(x_q))?;
```

Comment:
The constant doesn't have to be assigned in the same region or at the same offset; if it is, then it is likely that one fixed could will not be sufficient. There should be a utility to assign a constant at the next unused cell in that column.